ocaml-lsp-server/src/semantic_highlighting.ml

```diff
@@ -0,0 +1,827 @@
+    | Operator
+    | Decorator
+
+  let list =
+    [ (Module, "namespace")
+    ; (Type, "type")
+    ; (Class, "class")
+    ; (Enum, "enum")
+    ; (Module_type, "interface")
+    ; (Record, "struct")
+    ; (Type_parameter, "typeParameter")
+    ; (Parameter, "parameter")
+    ; (Variable, "variable")
+    ; (Property, "property")
+    ; (Enum_member, "enumMember")
+    ; (Event, "event")
+    ; (Function, "function")
+    ; (Method, "method")
+    ; (Macro, "macro")
+    ; (Keyword, "keyword")
+    ; (Modifier, "modifier")
+    ; (Comment, "comment")
+    ; (String, "string")
+    ; (Number, "number")
+    ; (Regexp, "regexp")
+    ; (Operator, "operator")
+    ; (Decorator, "decorator")
+    ]
+
+  let int_of_t (t : t) =
+    match t with
+    | Module -> 0 (* defined by index in [list] defined above *)
+    | Type -> 1
+    | Class -> 2
+    | Enum -> 3
+    | Module_type -> 4
+    | Record -> 5
+    | Type_parameter -> 6
+    | Parameter -> 7
+    | Variable -> 8
+    | Property -> 9
+    | Enum_member -> 10
+    | Event -> 11
+    | Function -> 12
+    | Method -> 13
+    | Macro -> 14
+    | Keyword -> 15
+    | Modifier -> 16
+    | Comment -> 17
+    | String -> 18
+    | Number -> 19
+    | Regexp -> 20
+    | Operator -> 21
+    | Decorator -> 22
+end [@warning "-37"]
+
+module Token_modifier = struct
+  type t =
+    | Declaration
+    | Definition
+    | Readonly
+    | Static
+    | Deprecated
+    | Abstract
+    | Async
+    | Modification
+    | Documentation
+    | Default_library
+
+  let list =
+    [ (Declaration, "declaration")
+    ; (Definition, "definition")
+    ; (Readonly, "readonly")
+    ; (Static, "static")
+    ; (Deprecated, "deprecated")
+    ; (Abstract, "abstract")
+    ; (Async, "async")
+    ; (Modification, "modification")
+    ; (Documentation, "documentation")
+    ; (Default_library, "defaultLibrary")
+    ]
+
+  let int_of_t (t : t) =
+    match t with
+    | Declaration -> 0 (* defined by index in [list] *)
+    | Definition -> 1
+    | Readonly -> 2
+    | Static -> 3
+    | Deprecated -> 4
+    | Abstract -> 5
+    | Async -> 6
+    | Modification -> 7
+    | Documentation -> 8
+    | Default_library -> 9
+
+  let int_of_ts (lst : t list) =
+    List.fold_left lst ~init:0 ~f:(fun i t -> i lor (1 lsl int_of_t t))
+end [@warning "-37"]
+
+let legend =
+  SemanticTokensLegend.create
+    ~tokenTypes:(List.map ~f:snd Token_type.list)
+    ~tokenModifiers:(List.map ~f:snd Token_modifier.list)
+
+module Tokens : sig
+  type t
+
+  val create : unit -> t
+
+  val append_token : t -> Loc.t -> Token_type.t -> Token_modifier.t list -> unit
+
+  val append_token' :
+       t
+    -> Position.t
+    -> length:int
+    -> Token_type.t
+    -> Token_modifier.t list
+    -> unit
+
+  val encode_tokens : t -> int array
+end = struct
+  type token =
+    { start : Position.t
+    ; length : int
+    ; token_type : int
+    ; token_modifiers : int
+    }
+
+  type t =
+    { mutable tokens : token list
```

Comment:
how big are these token lists in practice? also, can they have duplicate token types?

Reply:
> how big are these token lists in practice?

For `document_store.ml` in this PR, the `data` array is ~2575 integers long. This means there are 515 tokens. 

> can they have duplicate token types? 

If you really mean token types, then yes, because almost every symbol whether it's a variable name or a function name gets a token type, so duplication is inevitable. Duplications I guess do happen but I don't know whether we could do some sort of caching. 

One perf improvement perhaps would be to not use a list of token modifiers but use raw integers hidden by opaque type, e.g., something like `[Definition; Abstract]` could become `Token_modifiers.(definition % abstract)` which behind the scenes would simply do `lor`s

Reply:
> One perf improvement perhaps would be to not use a list of token modifiers but use raw integers hidden by opaque type, e.g., something like [Definition; Abstract] could become Token_modifiers.(definition % abstract) which behind the scenes would simply do lors

That sounds like the right way to go to me.